### ISSUE TITLE
Long background-write queue may lead to unnecessary hinted handoff

### ISSUE THREAD

**Issue by nyh** [scylladb/scylla#4191]
This issue is about "throttling", our existing mechanism of flow control of writes in regular tables. This mechanism involves measuring the size of "background" writes (user's write requests which have already achieved CL, and thereby replied, but writes to additional replicas haven't finished yet), and when this reaches a threshold (10% of memory), we do all further work in the foreground (see https://www.scylladb.com/2018/12/04/worry-free-ingestion-flow-control/ for more details).

This basically means that we fully expect this "10% of memory" to be full for extended periods of time (otherwise, throttling doesn't even kick in). At that point, when a coordinator has tens of thousands of unreplied replica writes, it means that one of the replicas has tens of thousands of requests in its queue. This means that a new write to this replica will wait a very long time until responded. This doesn't cause any problems for the user (the throttling algorithm correctly replies to the user as soon as any replica write finished), but does cause an internal problem: Internally, Scylla does replica writes with a 2-second (write_request_timeout_in_ms in scylla.yaml) timeout. When the long queue causes every write to one of the replicas to be delayed more than this timeout, Scylla wrongly believe each of these replica writes failed, and uses **hinted handoff**, i.e., writes these writes on disk to be replayed later.

Note that whether we run into these internal write timeouts and hinted handoffs depends on the size of memory per shard (10% of which is the queue size) and shard performance, so depending on the machine type and size, we may or may not see it happening. I consistently see it with a cluster of three i3.large machines (2 cpus, 8 GB memory per CPU), and cassandra-stress run with RF=3, CL=1 (`cassandra-stress write duration=5m no-warmup -node ...  -rate threads=100 -mode native cql3`).

The use of hinted handoff in this case is of course wasteful, because we treat an expected situation (when the queue is full, we know every new replica write will take a long time depending on the arbitrary queue size chosen, and this may as well be more than the arbitrary 2 second timeout) as if something is wrong.

While it is obvious that using hinted-handoff is wasteful, what is even stranger is that somehow it causes mysterious things to happen to Scylla's **metrics**. For example, in the above cassandra-stress run, cassandra-stress reported acceptable performance, but Scylla's metrics in prometheus showed this:
![a5](https://user-images.githubusercontent.com/584227/52267101-d8980d00-2940-11e9-878c-89cac0575aab.png)
![a6](https://user-images.githubusercontent.com/584227/52267107-dcc42a80-2940-11e9-8f6b-d03b63e64eaa.png)
while in reality, the client (cassandra-stress) showed no timeouts, and low tail latencies. Our **theory** is that the continuation responsible of accounting the end of the request somehow gets delayed for a very long time because of a lot of continuations on the queue, or something. It's very unclear though, why specifically these continuations have a problem, or why exactly the extra hinted handoff is what brings on these problems. This needs to be debugged and understood.

**Comment by gleb-cloudius** [comment]
I do not understand what throttling has to do with any of that. If no
throttling happens whatsoever background requests will fail and be
written to HH after configured write timeout. There is no point in
waiting for them any longer because they will be likely abandoned by
the replica by this point. Replicas knows to not bother with a request
for more than configured timeout as well. Again nothing here has to do
with throttling.

--
			Gleb.

**Comment by nyh** [comment]
@gleb-cloudius throttling is in charge of slowing down the client to exactly the rate at which we can **complete** requests. However, currently, it doesn't slow the client at all until the entire quota of 10% of memory is full, and only then starts to slow the client down. At that point, the client is at the right rate, but the "10% of memory" queue will always be full. If throttling started slowing down the client earlier, we wouldn't have this huge queue. So it's not that throttling is **causing** this problem, but rather that it is not preventing it.

You made an interesting point about replicas ignoring the request after the configured timeout. This means that it's not enough to disable hinted-handoff, we really need to ensure that the background-write queue isn't so large that it causes timeouts (we can also increase the configured timeout - there's nothing holy about two seconds). This again means the throttling algorithm needs to start slowing the client earlier - not when the queue reaches the arbitrary number "10% of memory".

And of course, all of this is unrelated to the question why, when hinted-handoff is used, we see the bizarre problems with the metrics.

**Comment by gleb-cloudius** [comment]
You lump a lot of unrelated thing into this issue. If this issue is
about throttling not always preventing overload? If yes than what HH has to
do with it? If this issue is about HH (seams to be) overloading CPU,
then what throttling has to do with it. I am confused what this issue is
about.

--
			Gleb.

**Comment by nyh** [comment]
This issue is about the bad things, namely HH and bizarre metrics, that happen once the background write queue is very long.

A separate question is whether throttling should allow the background write queue to become this long. *That* is not the point of this issue. I agree that the title I used is probably misleading. I'll change it.

**Comment by gleb-cloudius** [comment]
On Tue, Feb 05, 2019 at 03:45:22AM -0800, nyh wrote:
> This issue is about the bad things, namely HH and bizarre metrics, that happen once the background write queue is very long.
> 
OK. I see now.

--
			Gleb.

**Comment by psarna** [comment]
@nyh please verify if I understand this issue correctly:

If the number of unreplied write requests hits the memory limit on the coordinator side, it's expected that some replicas have a long queue on their side. In that case, when the request is timed out on the coordinator, we should not bother the coordinator with storing hints, because the replica is expected to process the request for a long time, right? If that's the case, I just wonder if this situation is distinguishable from a replica that gets overloaded with write requests and then dies. In such scenario, hints would work as anti-entropy as expected, while if we stopped storing hints for that replica, we would not resend some of this latest data when it comes back to life. So the question is whether storing hints for extended periods of time for a node with a full write queue is worth having slightly better insurance in case this node fails - and I'm not sure what's the right answer, because if users are not interested in this mechanism, they can just turn it off and repair the nodes slightly more often.

**Comment by nyh** [comment]
I did not suggest to disable hints. 
I think there are two related problems here that together cause the mess I saw:

1. Our "throttling" (background base writes) algorithm works, unfortunately, by creating a very long queue of writes on at least one of the replicas. If this queue is long enough that it takes more than the timeout (e.g., 2 seconds) for a request to go through the queue, all these requests time out. Which is already bad and wasteful - unrelated to hints.
2. After the above problem causes all writes to a particular replica to timeout, the coordinator (deliberately, and correctly) starts writing hints for this replica. In my experiments this caused weird performance problems (I don't know if this is still true, you'll have to check). Part of it is probably the understandable "avalanche" problem - we have here a replica which already has a huge queue of writes to do - and now the hints code start **replaying** even more writes to this node, again and again! But there may be additional bugs in the hints code, because I saw some truly weird things in the monitoring.

Regarding the hints (2), I want to note that the problem is probably not the fact that the coordinator stores hints, but rather that it replays them. It's a good idea to start replaying hints when a dead replica comes back up. But if the replica is still overloaded in the sense that it will timeout every request sent to it, it is pointless to start replaying hints to it.
But to be honest, I'm more worried that we got to the point where a replica has such a long queue of writes that every write to it will timeout. I think we need to solve that problem more than we need hints to work well in this scenario that shouldn't happen in the first place.

**Comment by psarna** [comment]
Right, but hints themselves are sent with a steady low concurrency (if I read the code correctly), so it probably still makes sense to try and replay them, because the node might eventually stop being overloaded. My point is that perhaps this issue is not really `Long background-write queue may lead to unnecessary hinted handoff`, but more like `Long background-write queue usually means that some replicas have long write queues, and that causes all writes to these replicas to fail`?

It feels like solving this issue might not be related to hinted handoff at all, because the root problem will also be observed with hinted handoff disabled - a replica effectively stops succeeding with write operations, because its internal queue gets so long that each operation just times out before it has a chance of being executed.

I'll try to replicate the scenario from the issue description and then, if I manage to see that no writes succeed on a replica, try to figure out why exactly that happens and how we can mitigate it.

**Comment by nyh** [comment]
> Right, but hints themselves are sent with a steady low concurrency (if I read the code correctly), so it probably still makes sense to try and replay them, because the node might eventually stop being overloaded.

Maybe this code changed in the year that has passed since I created this issue, because in my tests, I remember that enabling hints definitely made everything worse (and weird - such as those strange latency numbers which the client didn't see).

> My point is that perhaps this issue is not really `Long background-write queue may lead to unnecessary hinted handoff`, but more like `Long background-write queue usually means that some replicas have long write queues, and that causes all writes to these replicas to fail`?

This is true. At some point if you can reproduce some problem, feel free to close this issue and open a new one which more clearly focuses on a specific problem.
 
> It feels like solving this issue might not be related to hinted handoff at all, because the root problem will also be observed with hinted handoff disabled - a replica effectively stops succeeding with write operations, because its internal queue gets so long that each operation just times out before it has a chance of being executed.
> 
> I'll try to replicate the scenario from the issue description and then, if I manage to see that no writes succeed on a replica, try to figure out why exactly that happens and how we can mitigate it.

Good plan. Just note that to reproduce this issue, you'll need:
1. One node to be slightly (even very slightly) slower than the other two, so that CL<ALL writes will cause it to have more and more unfinished replica writes. You may want to artificially slow it down (maybe even by running adding a silly slowdown fiber to Scylla?) if there is no natural slow node.
2. The problem occurs if the queue length (10% of memory), replica write throughput, and timeout (2 seconds), fulfill a certain inequality (queuelength/throughput > timeout). You may need to play with some of these parameters to make this happen. Throughput is fixed (depending on your hardware) but the memory size (or 10% number), and the timeout, can be changed.

**Comment by psarna** [comment]
Alright, I can see timeouts happening on the slightly slower node, but not an overwhelming number, just a small percentage. And the overall throughput of the whole cluster drops (compared to the situation when the node is as fast as the other ones). I'll try modifying some parameters to reach a situation when almost all writes to that node end up failing.

By the way, as for slowing a single node, perf with sufficiently high frequency sampling seems to do a great job for simulating that - I get very consistent results indicating that the node is indeed slower than the rest:
```bash
sudo perf top -p $SCYLLA_PID -g -F 65536
```

**Comment by psarna** [comment]
Hm, I'm getting consistent results which suggest that throttling works as expected and doesn't cause anomalies on replicas... I tried multiple times with different loads, with the nodes 100% utilized in most of the cases, on a cluster of 3 i3.large instances. The loading process sends write requests with a fixed concurrency for 5 minutes.

The screenshot below shows two runs - one performed on a cluster with fully symmetrical replicas. The cluster is able to ingest a certain number of writes, and after the load finishes there's a big bump of hints, as if a large number of them has piled up - this is the part I didn't understand and I'll try to run a much longer test case tomorrow and see what happens. It's weird, because grafana does not indicate a large number of failed writes (<10/s), so where does this huge pile of hint sending operations come from?

In any case, the second part of the timeline shows the same cluster under the same load, but one of the node is considerably slower, due to the scylla process being monitored by perf with high frequency of events. We can observe that the overall throughput also decreases considerably, and the number of write timeouts on the slower node is a small percentage of successfully applied operations. And it suggests that throttling managed to successfully stop the client from overloading the cluster at the cost of a small percentage of write operations failing on the problematic node.

![bg_read_with_slower_node](https://user-images.githubusercontent.com/10433047/110805433-5c8bf400-8281-11eb-9f70-996d178e0a87.png)

So now, the question is what caused the hint storm in case of a perfectly symmetrical cluster, and would it potentially cause problems if the experiment lasted significantly more than 5 minutes - I'll try that tomorrow.

**Comment by psarna** [comment]
Oh and the most important thing to observe is that during the first part of the timeline, when the cluster is symmetrical, the number of written hints per second is really large - and that's the obvious direct cause of the spike in hints sent - but I don't yet get why so many hints are stored in the first place, since there are no observed failed writes...

**Comment by psarna** [comment]
The current working theory is that these hints are results of writes that were sent to a replica, but then they timed out on the coordinator. I assumed that they would be visible in the monitorings under "writes timed out", but maybe I messed something up and this particular metrics is replica-side... In any case, I do see *some* coordinator-side write timeouts in the logs, but the rate (~100/s) does not match the 40kops/s of storing hints.
![write_timeouts](https://user-images.githubusercontent.com/10433047/110810077-bb536c80-8285-11eb-916b-91b317d4b9dc.png)

**Comment by gleb-cloudius** [comment]
On Thu, Mar 11, 2021 at 07:17:07AM -0800, Piotr Sarna wrote:
> Oh and the most important thing to observe is that during the first part of the timeline, when the cluster is symmetrical, the number of written hints per second is really large - and that's the obvious direct cause of the spike in hints sent - but I don't yet get why so many hints are stored in the first place, since there are no observed failed writes...
> 

Because hints are not stored for writes that are failed?

--
			Gleb.

**Comment by psarna** [comment]
Yeah, I misinterpreted the metrics. I thought that a "failed write" meant a failed coordinator->replica operation, while it means a failed client<->coordinator write (as in, CL wasn't met).

**Comment by psarna** [comment]
so what I'm observing here is simply a client sending enough stuff to overload replicas with writes, which in turn causes the coordinator to store lots of hints. I suppose that if I ran the scenario long enough for hints manager to start dropping requests due to overload, I'll just observe OverloadedExceptions on the client side.

**Comment by psarna** [comment]
That's what I was missing (thanks @piodul), metrics for failed background replica writes to remote nodes (a.k.a. `$func(rate(scylla_storage_proxy_coordinator_background_replica_writes_failed_remote_node{instance=~"[[node]]",cluster=~"$cluster|$^", dc=~"$dc", shard=~"[[shard]]"}[1m])) by ([[by]])`). Not surprisingly, it corresponds 1:1 with the number of stored hints:
![background_replica_failed_writes](https://user-images.githubusercontent.com/10433047/110812630-0ff7e700-8288-11eb-9a01-46c0d10957b1.png)

**Comment by psarna** [comment]
I ran a longer scenario, in which I start with a fully symmetrical cluster, and in the middle of the scenario one of the node suddenly gets slower (via starting perf with high sampling rate on the scylla process).

The results do not indicate any issues - cluster's throughput is correctly reduced to handle the load, and the slower node receives lots of hints after the main load is gone, which only means that the hints served their purpose and were successfully resent.

What's curious is that while the number of failed replica writes (the ones initiated by the coordinator and sent to replicas) is quite high in a symmetrical cluster, it drops to near-zero when the cluster gets asymmetrical, which suggests that throttling does a better job if the difference between nodes is more substantial. The only symptom of overload I saw on the slower node is tons of log messages informing that a cross-shard semaphore timed out:
```
Mar 12 13:55:04 sarna-root-db-node-94a3493d-3 scylla: [shard 0] storage_proxy - Failed to apply mutation from 10.0.1.197#0: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#2 1->0 semaphore)
Mar 12 13:55:04 sarna-root-db-node-94a3493d-3 scylla: [shard 0] storage_proxy - Failed to apply mutation from 10.0.1.197#0: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#2 1->0 semaphore)
Mar 12 13:55:04 sarna-root-db-node-94a3493d-3 scylla: [shard 0] storage_proxy - Failed to apply mutation from 10.0.1.197#0: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#2 1->0 semaphore)
Mar 12 13:55:04 sarna-root-db-node-94a3493d-3 scylla: [shard 0] storage_proxy - Failed to apply mutation from 10.0.1.197#0: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#2 1->0 semaphore)
Mar 12 13:55:04 sarna-root-db-node-94a3493d-3 scylla: [shard 0] storage_proxy - Failed to apply mutation from 10.0.1.197#0: seastar::named_semaphore_timed_out (Semaphore timed out: smp_service_group#2 1->0 semaphore)
(... and so on, times billion)
```
, which may actually suggests that scylla-bench wasn't benefiting from shard awareness? Anyway, results from today's run:
![slow_in_the_middle](https://user-images.githubusercontent.com/10433047/110966296-67b05400-8355-11eb-80a9-5cd3f6780648.png)

The final spike in hints is the result of the main load ending, which creates the opportunity for other nodes to catch up on hints created for the slower one.

In any case, even though the metrics are quite educational for me, I didn't reproduce the problem mentioned in this issue so far.

**Comment by nyh** [comment]
@psarna I don't understand most of your comments above, but I liked the graph in https://github.com/scylladb/scylla/issues/4191#issuecomment-796825996 (with the hints written and hints sent), and have some comments about it:

First, the graph suggests that indeed the throttling is working, and indeed hints are working.
This specific graph doesn't show if there is any sign of the bug I saw with latencies. Maybe indeed the original bug I saw is gone.

But MORE IMPORTANTLY, although the behavior seen in this graph is easy to justify, and in many senses is "reasonable", is it the **desirable** thing that should happen? Would argue the answer is NO:
Your workload runs for a few minutes, and can just as well run for hours. It is a "good" and "gentle" batch workload with fixed concurrency - it is not "overloading" the cluster in any sense, and the best thing to do for this workload is to slow it down to the "right" speed, so it can continue to work normally for hours, without any errors. Scylla did slow down the workload, but not enough enough -  we allowed a backlog to accumulate so much (on the single slow now) that now Scylla is wasting a lot of the resources on storing hints, which end up slowing the workload themselves too (because this hints stuff takes resources). Had Scylla, instead, slowed down the workload a bit more, e.g., to fill a half-second-worth queue instead of a more-than-timeout queue, we would get higher throughput (because no CPU is wasted on hints) and lower disk waste (on hints).

ֵMoreover, as you also noted, after your workload will run for a long time (e.g., hours), the current code will either start replaying hints quickly (further reducing throughput!) or - I think more likely - will not replay hints fast enough, will run out of disk space for these hints and start to fail random requests, slowing down the workload even more because the client will just retry those requests and cause even more churn.

**Comment by gleb-cloudius** [comment]
On Sun, Mar 14, 2021 at 12:31:14AM -0800, nyh wrote:
> ֵMoreover, as you also noted, after your workload will run for a long time (e.g., hours), the current code will either start replaying hints quickly (further reducing throughput!) or - I think more likely - will not replay hints fast enough, will run out of disk space for these hints and start to fail random requests, slowing down the workload even more because the client will just retry those requests and cause even more churn.
> 
Running out of disk space for hints will not cause random requests to
fail.

--
			Gleb.

**Comment by nyh** [comment]
> Running out of disk space for hints will not cause random requests to fail.

You're right. I was thinking of running out of memory buffer (which causes a "overloaded exception").
Running out of disk space will simply drop hints. Whether or not we consider this a problem is questionable (after all, repair will fix it too).

I'm having second thoughts about what will happen in this workload. Let me try to explain what I think will happen in this workload and why I think it's bad:

Let's look at a numeric example: We have two replicas that can do 10,000 writes per second, and one that can do 9,000 writes per second, and we write with quorum. Initially, the workload does 10,000 writes per second, and the queue of background writes grows by 1,000 items each second. Our theory was that things continue as follows: After a while, this queue grows to the limit (10% memory), and "throttling" starts. At that point and only at that point, the workload slows down to 9,000 requests per second (what the slow node can do). But if 10% of the memory means that slow replica already has (for example) 90,000 items queued up, these items take 10 seconds to clear up (at 9,000 a second), so because they are processed today in FIFO order, all these queued write will timeout and will not happen. All 90,000 writes will be missing on this replica and will (if we did not run out of diskspace) be stored as hints on the coordinator.

That's a nice story, but I'm starting to think this is not what will actually happen. I think we'll never actually get to the 10% memory or do any throttling! I think that if say the timeout is 2 seconds, then as soon as we have more than 18,000 writes queued up on the slow replica, the first ones are already more than the timeout old, so it takes a tiny amount of time for the replica to realize this and skip those first requests - skipping a write does not have the same resource use as fully perfoming a write. So the queue at the replica will actually be kept around 18,000 writes, not 90,000 writes. Which means the queue at the coordinator will never reach 10% of memory, and official throttling will never happen at all; Instead the coordinator will be slowed down unofficially, by doing garbage work (writing hints).

**Comment by psarna** [comment]
I'm working half-day today, so I don't have any epiphanies to present, but I'll drop the results of a 3h run of the same test - with one node being slower than others. The results are definitely interesting to see. I haven't studied them yet, so no interpretations from my side.

Detailed view:
![4h_experiment_detailed](https://user-images.githubusercontent.com/10433047/111174987-3b404600-85a8-11eb-86a3-c6da1bad47e8.png)
Overview of the cluster:
![4h_exp_overview](https://user-images.githubusercontent.com/10433047/111175009-3f6c6380-85a8-11eb-9f6a-7105f96f9b6a.png)
Overview of just the overloaded node:
![4h_exp_overview_overloaded_only](https://user-images.githubusercontent.com/10433047/111175021-41cebd80-85a8-11eb-8969-eef81bc40f9b.png)

**Comment by psarna** [comment]
One more panel from yesterday - at some point, the overloaded node cannot store more hints and starts dropping them, which is clearly visible and correlates with the number of timed out writes on that node:
![4h_exp_hints_dropped](https://user-images.githubusercontent.com/10433047/111270545-49d24000-8630-11eb-8853-9d35bf3d2166.png)
